### PR TITLE
Add: ecvrf coordinator integration

### PR DESCRIPTION
**Pull request by Prajjawalk**
@ProgramCrafter I have added couple of changes in `code.func` file, and added some tests in `tests/lottery.spec.ts`.

To run test - 
npx mocha -v --require ts-node/register tests/lottery.spec.ts 

Telegram Id - https://t.me/prajjawal003

### CODE REVIEW

**Comment by ProgramCrafter** [comment]
Hello! Have you also recompiled the code after all changes?

File: func/code.func [hunk 1]
```diff
@@ -16,6 +16,9 @@
 int workchains_equal?(slice addr1, slice addr2) asm "REWRITESTDADDR DROP SWAP REWRITESTDADDR DROP EQUAL";
 int slices_equal?(slice s1, slice s2) asm "SDEQ";
 int my_tons() asm "BALANCE" "FIRST";
+;; slice my_address() asm "MYADDR";
+global int rand_no;
+global int rand_updated;
```
Comment (ProgramCrafter) [inline func/code.func]: Global variables aren't automatically saved; you need to pack them in `pack_state` and restore in `unpack_state`.

File: func/code.func [hunk 2]
```diff
@@ -40,7 +43,7 @@ _ pack_state(int o_seqno, int n_wins, int in_amount, int out_amount, int hash, s
       begin_cell()                ;; SECOND CELL:
         .store_slice(owner1)      ;; 267 bit | owner 1, @sergei_ton
```
Comment (ProgramCrafter) [inline func/code.func]: I think only one owner should be left; if needed, it may be a contract forwarding winnings to other wallets.

File: func/code.func [hunk 3]
```diff
@@ -92,73 +95,90 @@ _ unpack_order(slice order) inline_ref {
 
   slice sender = cs~load_msg_addr();
 
-  (slice stats, cell orders, slice owner1, slice owner2, slice echo) = unpack_state();
+  (slice stats, cell orders, slice owner1, slice owner2, slice ecvrf) = unpack_state();
+
+  ;; subscribe to ecvrf
```
Comment (ProgramCrafter) [inline func/code.func]: ECVRF subscription is indeed required so lottery can function automatically.

File: func/code.func [hunk 4]
```diff
@@ -92,73 +95,90 @@ _ unpack_order(slice order) inline_ref {
...
+    (int o_seqno, int n_wins, int in_amount, int out_amount, int hash) = unpack_stats(stats);
+    ;; Calculating new hash
+    hash = cell_hash(begin_cell()
+      .store_uint(hash, 256)
+      .store_uint(msg_hash, 256)
+      .store_uint(cur_lt(), 64)
+      .store_uint(now(), 64)
+      .end_cell()
+    );
+    randomize(hash);
```
Comment (ProgramCrafter) [inline func/code.func]: `randomize()`, if needed, should be done before L138 `rand_no = rand(10);`

File: func/code.func [hunk 5]
```diff
@@ -92,73 +95,90 @@ _ unpack_order(slice order) inline_ref {
...
+    if (rand_no >= 4) {
```
Comment (ProgramCrafter) [inline func/code.func:116]: There is an issue since player can know random number stored in the contract and thus knows whether the bet is winning or losing.

File: func/code.func [hunk 6]
```diff
@@ -213,3 +233,7 @@ int get_pubkey() method_id {
 int balance() method_id {
     return my_tons();
 }
+
+int get_random() method_id {
+  return rand_no;
```
Comment (ProgramCrafter) [inline func/code.func]: Should `unpack_data()` first, since global variables don't persist between transactions.

**Comment by Prajjawalk** [comment]
yep, I have recompiled and rebuilt the code after changes using toncli

File: tests/lottery.spec.ts [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    async getLatestRand(provider: ContractProvider): Promise<number> {
+        return (await provider.get('get_random', [])).stack.readNumber()
+    }
+}
+
+describe("lottery test", () => {
+    let blockchain: Blockchain;
+    let ecvrf: SandboxContract<CoordinatorUnit>;
+    let lottery: SandboxContract<LotteryUint>;
+    const secretEcvrf = BigInt(123456);
```
Comment (ProgramCrafter) [inline tests/lottery.spec.ts]: You can use `123456n` instead of `BigInt(123456)`. Also applies to all further cases.

File: func/code.func [hunk 8]
```diff
@@ -16,6 +16,9 @@
 int workchains_equal?(slice addr1, slice addr2) asm "REWRITESTDADDR DROP SWAP REWRITESTDADDR DROP EQUAL";
```
Comment (ProgramCrafter) [inline func/code.func:16]: Special thanks could be removed if the hashing algorithm is replaced with ECVRF. Also, you can add yourself as a third author.

**Review by ProgramCrafter — COMMENTED** [review]
Here are flaws that I've noticed on a quick glance

**Comment by Prajjawalk** [comment]
Actually I am trying to return global variable here - https://github.com/Prajjawalk/ton-lottery-ecvrf/blob/d710ea820720f2f422e8bdc944e965add05dc2d8/func/code.func#L238. Does this cause any issue?

File: func/code.func (hunk 2 above)
Comment (ProgramCrafter) [reply func/code.func]: You can use any owner address during development, of course; no need to include splitter contract in this repository.

File: func/code.func (hunk 5 above)
Comment (Prajjawalk) [reply func/code.func:116]: Actually I was thinking the get_random function as a way of testing if random numbers are actually being updated. Could there be any other way of testing?

**Comment by Prajjawalk** [comment]
@ProgramCrafter I have added the changes you suggested, have kept get_random() function for debugging purpose which I will remove later.

**Comment by ProgramCrafter** [comment]
https://github.com/ProgramCrafter/ton-lottery-ecvrf/pull/2#discussion_r1372992305
> There is an issue since player can know random number stored in the contract and thus knows whether the bet is winning or losing.

By this, I mean an algorithm flaw of the current lottery. It first generates random value (and conveniently stores it) and then accepts bets, meaning that outcome is predetermined, so user can send a few TON if current number is losing (to request new number), and otherwise send a few thousand TON.

The lottery can:
- a) send one message to ECVRF coordinator *after* each bet, upon reply send winnings to user;
- b) use both ECVRF-generated random number and lt of incoming message, hashing them together, in hope that updating numbers sufficiently quickly won't allow users to determine outcome of their bet.

File: tests/lottery.spec.ts [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+        await ecvrf.sendSubscribeRandom(deployer.getSender(), 610000000n, lottery.address);
+    })
+
+    it("should update random number for coordinator", async () => {
+        const alpha = await ecvrf.getAlpha();
+        const pi = await ecvrf.getCalcPiFromAlpha(secretEcvrf, alpha);
+        await ecvrf.sendProvideRandomness(pi, keyReplay.secretKey);
+        let random_1 = await lottery.getLatestRand();
+        await ecvrf.sendProvideRandomness(pi, keyReplay.secretKey);
```
Comment (ProgramCrafter) [inline tests/lottery.spec.ts]: If I implemented coordinator contract correctly, this won't work: alpha and pi should be updated on each outgoing message (they are, essentially, seed and random number itself).

File: tests/lottery.spec.ts [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        });
+
+        await player1.send({
+            to: lottery.address,
+            value: initialBet * 2n,
+            sendMode: SendMode.PAY_GAS_SEPARATELY
+        })
+
+        const player1FinalBalance = await player1.getBalance();
+        expect(player1FinalBalance - player1IntermediateBalance).to.be.equal(initialBet) //(initialBet * 4);
```
Comment (ProgramCrafter) [inline tests/lottery.spec.ts]: May be better captured with snapshots?

**Comment by ProgramCrafter** [comment]
> have kept get_random() function for debugging purpose which I will remove later

Actually, even if it's removed the problem will remain since anyone can download complete state of contract from the blockchain.

File: func/code.func [hunk 11]
```diff
@@ -92,105 +93,120 @@ _ unpack_order(slice order) inline_ref {
...
-    .store_coins(0)
-    .store_uint(0, 1 + 4 + 4 + 64 + 32 + 1 + 1)
-    .store_slice(sender)
-    .store_coins(msg_value);
-  send_raw_message(msg.end_cell(), 64);
+  (slice player, int o_amount) = (in_msg_body~load_msg_addr(), in_msg_body~load_coins());
+  ;; lottery inactive if balance < 2 stake + 1 TON
+  if (my_tons() < o_amount * 2 + 1000000000) {
+      send_tons(player, msg_value, 2);
+  } else {
```
Comment (ProgramCrafter) [inline func/code.func]: Would be cleaner to `return ();` above?

File: func/code.func (hunk 3 above)
Comment (ProgramCrafter) [reply func/code.func]: Although probably not in this place, since now it's called even on owner's deposit or reception of random number.

File: func/code.func [hunk 12]
```diff
@@ -92,105 +93,120 @@ _ unpack_order(slice order) inline_ref {
...
   slice sender = cs~load_msg_addr();
 
-  (slice stats, cell orders, slice owner1, slice owner2, slice echo) = unpack_state();
+  (slice stats, cell orders, slice owner, slice ecvrf) = unpack_state();
+  ;; subscribe to ecvrf 
+  send_raw_message(
+    begin_cell()
+    .store_uint(0x18, 6)
+    .store_slice(ecvrf)
+    .store_coins(0)
```
Comment (ProgramCrafter) [inline func/code.func]: Attempt to send 0 TON + gas fees will only send message, but not let the coordinator contract process it and add the lottery contract into broadcast group.

**Review by ProgramCrafter — CHANGES_REQUESTED** [review]

**Comment by Prajjawalk** [comment]
> > have kept get_random() function for debugging purpose which I will remove later
> 
> Actually, even if it's removed the problem will remain since anyone can download complete state of contract from the blockchain.

The problem is the contract either receives message for ECVRF coordinator or from player at a time. Is there any way to do this process synchronously in FunC- take bet from user -> send message to ECVRF -> receive updated random number -> utilize the random number for current user betting?

**Comment by ProgramCrafter** [comment]
No, there are no async-await-like structures in TVM and, consequently, FunC. The existing bet info should be saved.

I think it is acceptable to lock the lottery contract until it receives the response.

**Comment by Prajjawalk** [comment]
> No, there are no async-await-like structures in TVM and, consequently, FunC. The existing bet info should be saved.
> 
> I think it is acceptable to lock the lottery contract until it receives the response.

In this case I will have to make sure that new user gets to bet only if the random number is updated and lock the lottery contract until then, and random_updated flag will help now.

**Comment by ProgramCrafter** [comment]
> new user gets to bet only if the random number is updated

Understanding test: what can happen if the random number is updated, meaning that new user *already knows* the random number?

**Comment by Prajjawalk** [comment]
> > new user gets to bet only if the random number is updated
> 
> Understanding test: what can happen if the random number is updated, meaning that new user _already knows_ the random number?

Nope, will use a flag that will toggle once the player bets and toggles again once the new random number is received and the lottery of the player is decided. In this time period whoever tries to bet will get rejected.

**Comment by Prajjawalk** [comment]
@ProgramCrafter I have made couple of changes to contract, can you plz have a look

File: func/code.func [hunk 13]
```diff
@@ -90,107 +92,127 @@ _ unpack_order(slice order) inline_ref {
...
+
+  ;; new random number received from vrf
+  if (op == 0x069CECA8 & slices_equal?(sender, ecvrf)) {
+    int vrfrand = in_msg_body~load_uint(256);
+    (int o_seqno, int n_wins, int in_amount, int out_amount, int hash, int rand_updated) = unpack_stats(stats);
+    int rand_no = vrfrand % 10;
+    rand_updated = 1;
+
+    ;; Calculating new hash
+    hash = cell_hash(begin_cell()
```
Comment (ProgramCrafter) [inline func/code.func]: I think hash is no longer used?

File: func/code.func [hunk 14]
```diff
@@ -28,34 +28,36 @@ cell remove_old_orders(cell orders, int o_seqno) {
...
 
-_ pack_state(int o_seqno, int n_wins, int in_amount, int out_amount, int hash, slice owner1, slice owner2, slice echo, cell orders) inline {
+_ pack_state(int o_seqno, int n_wins, int in_amount, int out_amount, int hash, slice owner, slice ecvrf, slice player, int rand_updated, int player_coins, cell orders) inline {
   return begin_cell()             ;; FIRST CELL
     .store_uint(o_seqno, 64)      ;; 64 bit  | order seqno
     .store_uint(n_wins, 64)       ;; 64 bit  | number of wins
     .store_uint(in_amount, 256)   ;; 256 bit | ingoing amount
     .store_uint(out_amount, 256)  ;; 256 bit | outgoing amount
     .store_uint(hash, 256)        ;; 256 bit | seed salt
+    .store_uint(rand_updated, 32) ;; 32 bit  | bool rand_updated
```
Comment (ProgramCrafter) [inline func/code.func]: Can be replaced with `.store_bit(rand_updated)`, if `rand_updated` is always 0/-1.

Should also be renamed to `random_locked`.

**Review by ProgramCrafter — APPROVED** [review]
The contract turns out to be a good thingy)

File: func/code.func (hunk 14 above)
Comment (Prajjawalk) [reply func/code.func]: I couldn't find anything for store_bit, can you plz point similar function in doc?

Comment (ProgramCrafter) [reply func/code.func]: I may be mistaken but I certainly remember that this function exists. It may be embedded in FunC compiler, actually.

Comment (Prajjawalk) [reply func/code.func]: Actually -1 takes 257 bits so we can have boolean as 0/1 with 8 bit int size

Comment (ProgramCrafter) [reply func/code.func]: > Actually -1 takes 257 bits

It actually can be stored in smaller storage, and 1 bit is sufficient.
0/1 -> `.store_uint(random_locked, 1)`
0/-1 -> `.store_int(random_locked, 1)`, equivalent to `.store_bit(random_locked)`.

**Comment by Prajjawalk** [comment]
@ProgramCrafter I have made the recommended changes. In the unit test this line - https://github.com/Prajjawalk/ton-lottery-ecvrf/blob/7ce91988d7b57622f31d12731029bbf6552400e0/tests/lottery.spec.ts#L207 the transaction fails with exit code 9. Here I am just sending random number data to lottery contract, can you please check if there is anything I am missing in transaction structure.

File: func/code.func [hunk 15]
```diff
@@ -90,107 +92,127 @@ _ unpack_order(slice order) inline_ref {
...
-
-        in_amount += o_amount;
-        out_amount += o_amount * 2;
-        n_wins += 1;
-
-        send_tons(player, 2 * o_amount, 3);
+  int op = in_msg_body~load_uint(32);
+
+  ;; new random number received from vrf
+  if (op == 0x069CECA8 & slices_equal?(sender, ecvrf)) {
```
Comment (ProgramCrafter) [inline func/code.func]: Can have an issue with operators priority if & is executed before ==.

Comment (Prajjawalk) [reply func/code.func]: Will changing it to this help? -` if ((op == 0x069CECA8) & (slices_equal?(sender, ecvrf)))`

File: func/code.func [hunk 16]
```diff
@@ -28,34 +28,36 @@ cell remove_old_orders(cell orders, int o_seqno) {
...
+        .store_slice(player)      ;; 267 bit | current player
+        .store_coins(player_coins)
         .end_cell())
     .end_cell();
 }
 
 _ unpack_state() inline {
   var ds = get_data().begin_parse();
-  slice stats = ds~load_bits(896);
+  slice stats = ds~load_bits(896 + 64 + 32);
```
Comment (ProgramCrafter) [inline func/code.func]: The problem raising exit code 9 is here. There should be 897 bits loaded.

Comment (Prajjawalk) [reply func/code.func]: ah! Yes should be 897 bits, let me try again with the changes

**Comment by Prajjawalk** [comment]
@ProgramCrafter the coordinator contract - https://github.com/ProgramCrafter/ecvrf-coordinator-contract/blob/master/main.ts gives error code 6 when I try to run ecvrf related get methods in updated ton sandbox. 

package.json - 
```{
  "dependencies": {
    "@orbs-network/ton-access": "^2.3.3",
    "@ton-community/func-js": "^0.6.3-tvmbeta.1",
    "@ton-community/func-js-bin": "^0.4.5-tvmbeta.1",
    "@ton/core": "^0.53.0",
    "@ton/crypto": "^3.2.0",
    "@ton/sandbox": "^0.13.1",
    "@ton/test-utils": "^0.3.1",
    "@ton/ton": "^13.9.0",
    "dotenv": "^16.1.4",
    "ts-node": "^10.9.1",
    "yargs": "^17.7.2"
  },
  "devDependencies": {
    "@types/node": "^20.2.6",
    "prando": "^6.0.1"
  }
}
```
error log - 
```
throw new GetMethodError(res.output.vm_exit_code, BigInt(res.output.gas_used), res.logs, res.output.vm_log, res.debugLogs);
                  ^
GetMethodError: Unable to execute get method. Got exit_code: 6
    at SmartContract.get (/Users/prajjawalkhandelwal/Prajjawal/ecvrf-coordinator-contract/node_modules/@ton/sandbox/dist/blockchain/SmartContract.js:266:19)
    at Blockchain.runGetMethod (/Users/prajjawalkhandelwal/Prajjawal/ecvrf-coordinator-contract/node_modules/@ton/sandbox/dist/blockchain/Blockchain.js:108:50)
    at async BlockchainContractProvider.get (/Users/prajjawalkhandelwal/Prajjawal/ecvrf-coordinator-contract/node_modules/@ton/sandbox/dist/blockchain/BlockchainContractProvider.js:53:24)
    at async CoordinatorUnit.getCalcPiFromAlpha (/Users/prajjawalkhandelwal/Prajjawal/ecvrf-coordinator-contract/main.ts:72:28)
    at async /Users/prajjawalkhandelwal/Prajjawal/ecvrf-coordinator-contract/main.ts:133:13 {
  exitCode: 6,
  gasUsed: 664n,
  blockchainLogs: '[ 4][t 0][2023-10-31 09:50:36.031000][mc-config.cpp:260]\tneedSpecialSmc flag set\n' +
    '[ 4][t 0][2023-10-31 09:50:36.031000][mc-config.cpp:266]\tsmc dictionary created\n' +
    '[ 4][t 0][2023-10-31 09:50:36.031000][mc-config.cpp:469]\tworkchain description dictionary created\n' +
    '[ 4][t 0][2023-10-31 09:50:36.031000][mc-config.cpp:2040]\tunpacked info for workchain 0: basic=true, active=true, accept_msgs=true, min_split=0, max_split=4\n' +
    '[ 4][t 0][2023-10-31 09:50:36.031000][SmartContract.cpp:202]\tVM stack:\n' +
    ' [ 123456 CS{Cell{0042006575099f60e0dc01c901fabbee4f64d8b426bb624a82560c948aced0d0b088f1} bits: 0..264; refs: 0..0} 77421 ] \n' +
    '[ 4][t 0][2023-10-31 09:50:36.032000][SmartContract.cpp:223]\tVM log\n' +
    'execute SETCP 0\n' +
    'execute DICTPUSHCONST 19 (xC_,1)\n' +
    'execute DICTIGETJMPZ\n' +
    'execute OVER\n' +
    'handling exception code 6: invalid opcode\n' +
    'default exception handler, terminating vm with exit code 6\n' +
    '[ 4][t 0][2023-10-31 09:50:36.032000][SmartContract.cpp:226]\tVM stack:\n' +
    ' [ 0 ] \n' +
    '[ 4][t 0][2023-10-31 09:50:36.032000][SmartContract.cpp:227]\tVM exit code: 6\n' +
    '[ 4][t 0][2023-10-31 09:50:36.032000][SmartContract.cpp:228]\tVM accepted: true\n' +
    '[ 4][t 0][2023-10-31 09:50:36.032000][SmartContract.cpp:229]\tVM success: false\n',
  vmLogs: 'execute SETCP 0\n' +
    'execute DICTPUSHCONST 19 (xC_,1)\n' +
    'execute DICTIGETJMPZ\n' +
    'execute OVER\n' +
    'handling exception code 6: invalid opcode\n' +
    'default exception handler, terminating vm with exit code 6\n',
  debugLogs: ''
}
```

**Comment by ProgramCrafter** [comment]
That's why there is specific sandbox version to use. Coordinator contract
is based on opcodes that work in testnet only.

вт, 31 окт. 2023 г., 14:55 Prajjawalk ***@***.***>:

**Comment by Prajjawalk** [comment]
So in sandbox version `0.12.0-tvmbeta.1` did it ever happen that `from` field in transaction is undefined? Because it happened when I tried to test the transactions and as a result the `sender` value becomes undefined, eg -
```
AssertionError: Expected [
  {
    from: undefined,
    to: EQBveYtTq8QbIKHvPGKJdhRvd68cUnqkFwJtW6Q4kGULrzhs,
    on: EQBveYtTq8QbIKHvPGKJdhRvd68cUnqkFwJtW6Q4kGULrzhs,
    value: undefined,
    body: x{4B75C35F3D32744FEDE46F1221019E3B575E3911EFA717B51125EF7318FC841BC_}
     x{D001}
      x{6200091D4BB6FFE03DD75745947B3C219DD88C4C5C0887D46B6F301811778BF355CF200E4E1C0000000000000000000000000000069CECA800000000000000000000000000000000000000000000000000000000000000AF},
    inMessageBounced: undefined,
    inMessageBounceable: undefined,
    op: 1266008927,
    initData: undefined,
    initCode: undefined,
    deploy: false,
    lt: 20000000n,
    now: 1698748159,
    outMessagesCount: 1,
    oldStatus: 'active',
    endStatus: 'active',
    totalFees: 4190328n,
    aborted: false,
    destroyed: false,
    exitCode: 0,
    actionResultCode: 0,
    success: true
  },
  {
    from: undefined,
    to: EQASOpdt_8B7rq6LKPZ4QzuxGJi4EQ-o1t5gMCLvF-arnh55,
    on: EQASOpdt_8B7rq6LKPZ4QzuxGJi4EQ-o1t5gMCLvF-arnh55,
    value: 30000000n,
    body: x{069CECA800000000000000000000000000000000000000000000000000000000000000AF},
    inMessageBounced: false,
    inMessageBounceable: true,
    op: 110947496,
    initData: undefined,
    initCode: undefined,
    deploy: false,
    lt: 21000000n,
    now: 1698748159,
    outMessagesCount: 0,
    oldStatus: 'active',
    endStatus: 'active',
    totalFees: 1658000n,
    aborted: false,
    destroyed: false,
    exitCode: 0,
    actionResultCode: 0,
    success: true
  }
] to contain a transaction that matches pattern {
  from: EQBveYtTq8QbIKHvPGKJdhRvd68cUnqkFwJtW6Q4kGULrzhs,
  success: true
}
```
which I got from - 
```
let txRes = await mockEcvrf.send({
            to: lottery.address,
            value: 30000000n,
            body: beginCell().storeUint(0x069CECA8, 32).storeUint(rnd, 256).endCell(),
            sendMode: SendMode.PAY_GAS_SEPARATELY
        })
        expect(txRes.transactions).to.have.transaction({
            from: mockEcvrf.address,
            success: true
        })
```

**Comment by ProgramCrafter** [comment]
Do the tests work now?

**Comment by Prajjawalk** [comment]
> Do the tests work now?

Yes @ProgramCrafter they are working

**Comment by ProgramCrafter** [comment]
Thank you for the contribution!
